File: src/nuitrack_bridge/modules/user_tracker.cc
#include "nuitrack_bridge/modules/user_tracker.h"
#include "nuitrack/modules/UserTracker.h"
#include "nuitrack/types/UserFrame.h"
#include "nuitrack/Nuitrack.h"

#include <stdexcept>
#include <string>
#include <functional>

namespace nuitrack_bridge::user_tracker {

    // --- Helper Functions for Error Formatting ---
    static std::string format_nuitrack_error(const std::string& function_name, const std::string& nuitrack_error_what) {
        return "Nuitrack " + function_name + " failed: " + nuitrack_error_what;
    }
    static std::string format_std_error(const std::string& function_name, const std::string& std_error_what) {
        return "Standard exception in Nuitrack " + function_name + ": " + std_error_what;
    }
    static std::string format_unknown_error(const std::string& function_name) {
        return "Unknown exception during Nuitrack " + function_name;
    }

    // --- Bridge Function Implementations ---
    std::shared_ptr<UserTracker> createUserTracker() {
        try {
            return tdv::nuitrack::UserTracker::create();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::create", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::create", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::create"));
        }
    }

    uint64_t connectOnUpdateForAsync(const std::shared_ptr<UserTracker>& tracker, void* userFrameSender) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in connectOnUpdateForAsync"); }
        try {
            return tracker->connectOnUpdate(
                [userFrameSender](UserFrame::Ptr frame) {
                    rust_user_tracker_user_frame_dispatcher_async(frame, userFrameSender);
                });
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::connectOnUpdate", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::connectOnUpdate", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::connectOnUpdate"));
        }
    }

    void disconnectOnUpdate(const std::shared_ptr<UserTracker>& tracker, uint64_t handlerId) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in disconnectOnUpdate"); }
        try {
            tracker->disconnectOnUpdate(handlerId);
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::disconnectOnUpdate", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::disconnectOnUpdate", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::disconnectOnUpdate"));
        }
    }

    uint64_t connectOnNewUserForAsync(const std::shared_ptr<UserTracker>& tracker, void* newUserEventSender) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in connectOnNewUserForAsync"); }
        try {
            return tracker->connectOnNewUser(
                [newUserEventSender](int userID) {
                    rust_user_tracker_new_user_event_dispatcher_async(userID, newUserEventSender);
                });
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::connectOnNewUser", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::connectOnNewUser", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::connectOnNewUser"));
        }
    }

    void disconnectOnNewUser(const std::shared_ptr<UserTracker>& tracker, uint64_t handlerId) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in disconnectOnNewUser"); }
        try {
            tracker->disconnectOnNewUser(handlerId);
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::disconnectOnNewUser", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::disconnectOnNewUser", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::disconnectOnNewUser"));
        }
    }

    uint64_t connectOnLostUserForAsync(const std::shared_ptr<UserTracker>& tracker, void* lostUserEventSender) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in connectOnLostUserForAsync"); }
        try {
            return tracker->connectOnLostUser(
                [lostUserEventSender](int userID) {
                    rust_user_tracker_lost_user_event_dispatcher_async(userID, lostUserEventSender);
                });
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::connectOnLostUser", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::connectOnLostUser", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::connectOnLostUser"));
        }
    }

    void disconnectOnLostUser(const std::shared_ptr<UserTracker>& tracker, uint64_t handlerId) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in disconnectOnLostUser"); }
        try {
            tracker->disconnectOnLostUser(handlerId);
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::disconnectOnLostUser", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::disconnectOnLostUser", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::disconnectOnLostUser"));
        }
    }

    std::shared_ptr<UserFrame> getUserFrame(const std::shared_ptr<UserTracker>& tracker) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in getUserFrame"); }
        try {
            return tracker->getUserFrame();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::getUserFrame", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::getUserFrame", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::getUserFrame"));
        }
    }
    
    float getProcessingTime(const std::shared_ptr<UserTracker>& tracker) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in getProcessingTime"); }
        try {
            return tracker->getProcessingTime();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::getProcessingTime", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::getProcessingTime", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::getProcessingTime"));
        }
    }

    uint64_t getTrackerTimestamp(const std::shared_ptr<UserTracker>& tracker) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in getTrackerTimestamp"); }
        try {
            return tracker->getTimestamp();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::getTimestamp", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::getTimestamp", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::getTimestamp"));
        }
    }
    
    bool canUpdate(const std::shared_ptr<UserTracker>& tracker) {
        if (!tracker) { throw std::runtime_error("UserTracker instance is null in canUpdate"); }
        try {
            return tracker->canUpdate();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(format_nuitrack_error("UserTracker::canUpdate", e.what()));
        } catch (const std::exception& e) {
            throw std::runtime_error(format_std_error("UserTracker::canUpdate", e.what()));
        } catch (...) {
            throw std::runtime_error(format_unknown_error("UserTracker::canUpdate"));
        }
    }

}

File: src/nuitrack_bridge/modules/user_tracker.rs
#[cxx::bridge(namespace = "nuitrack_bridge::user_tracker")]
pub mod ffi {

    // Opaque C++ types that Rust will interact with via smart pointers.
    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        type UserTracker;
        type UserFrame = crate::nuitrack_bridge::types::user_frame::ffi::UserFrame;
    }

    unsafe extern "C++" {
        include!("nuitrack_bridge/modules/user_tracker.h");

        // We need a C++ type for the void* pointer.
        // This is aliased in `user_tracker.h` but must be declared here for cxx.
        pub type c_void;

        // --- Module Lifecycle ---
        #[cxx_name = "createUserTracker"]
        pub fn create_user_tracker() -> Result<SharedPtr<UserTracker>>;

        // --- On-Update Callbacks ---
        #[cxx_name = "connectOnUpdateForAsync"]
        pub unsafe fn connect_on_user_frame_async(
            tracker: &SharedPtr<UserTracker>,
            user_frame_sender: *mut c_void,
        ) -> Result<u64>;

        #[cxx_name = "disconnectOnUpdate"]
        pub fn disconnect_on_user_frame(
            tracker: &SharedPtr<UserTracker>,
            handler_id: u64,
        ) -> Result<()>;

        // --- New User Callbacks ---
        #[cxx_name = "connectOnNewUserForAsync"]
        pub unsafe fn connect_on_new_user_event_async(
            tracker: &SharedPtr<UserTracker>,
            new_user_event_sender: *mut c_void,
        ) -> Result<u64>;

        #[cxx_name = "disconnectOnNewUser"]
        pub fn disconnect_on_new_user_event(
            tracker: &SharedPtr<UserTracker>,
            handler_id: u64,
        ) -> Result<()>;

        // --- Lost User Callbacks ---
        #[cxx_name = "connectOnLostUserForAsync"]
        pub unsafe fn connect_on_lost_user_event_async(
            tracker: &SharedPtr<UserTracker>,
            lost_user_event_sender: *mut c_void,
        ) -> Result<u64>;

        #[cxx_name = "disconnectOnLostUser"]
        pub fn disconnect_on_lost_user_event(
            tracker: &SharedPtr<UserTracker>,
            handler_id: u64,
        ) -> Result<()>;

        // --- Synchronous Data Access ---
        #[cxx_name = "getUserFrame"]
        pub fn user_frame(tracker: &SharedPtr<UserTracker>) -> Result<SharedPtr<UserFrame>>;

        // --- Module Information ---
        #[cxx_name = "getProcessingTime"]
        pub fn processing_time(tracker: &SharedPtr<UserTracker>) -> Result<f32>;

        #[cxx_name = "getTrackerTimestamp"]
        pub fn tracker_timestamp(tracker: &SharedPtr<UserTracker>) -> Result<u64>;

        #[cxx_name = "canUpdate"]
        pub fn can_update(tracker: &SharedPtr<UserTracker>) -> Result<bool>;
    }
}

// Mark the opaque C++ type as safe to send across threads.
// This is an assertion that std::shared_ptr<UserTracker> is thread-safe.
unsafe impl Send for ffi::UserTracker {}
unsafe impl Sync for ffi::UserTracker {}